tokio/src/sync/mpsc/chan.rs

```diff
@@ -149,6 +168,50 @@ impl<T, S> Tx<T, S> {
     }
 }
 
+impl<T, S> TxWeak<T, S> {
+    fn new(inner: Weak<Chan<T, S>>) -> Self {
+        TxWeak { inner }
+    }
+
+    pub(super) fn upgrade(&self) -> Option<Tx<T, S>> {
+        let inner = self.inner.upgrade();
+
+        if let Some(inner) = inner {
+            // If we were able to upgrade, `Chan` is guaranteed to still exist,
+            // even though the channel might have been closed in the meantime.
+            // Need to check here whether the channel was actually closed.
+
+            let mut tx_count = inner.tx_count.load(Relaxed);
+            loop {
+                // FIXME Haven't thought the orderings on the CAS through yet
+                match inner
+                    .tx_count
+                    .compare_exchange(tx_count, tx_count + 1, SeqCst, SeqCst)
```

Comment:
I don't think you should increment the counter in the first place if it is already closed.

Reply:
Not entirely sure what you mean here. Do you think the Cas is not necessary here or do you want to check whether `tx_count` is 0 before going into the Cas?

Reply:
The CAS is only necessary if you check whether `tx_count` is 0 first.